Pasteboard history should not include items copied from ephemeral sessions.

megangardner opened the pull request:
#### 978bbafb7a3ab151703ba7e9694b78e2d63c3fe1
<pre>
Pasteboard history should not include items copied from ephemeral sessions.
<a href="https://bugs.webkit.org/show_bug.cgi?id=295442">https://bugs.webkit.org/show_bug.cgi?id=295442</a>
<a href="https://rdar.apple.com/153315024">rdar://153315024</a>

Reviewed by Abrar Rahman Protyasha.

In order to keep items copied in an ephemeral
session from showing up in pasteboard history,
we need to set a value before writing to the pasteboard.
We also have decided that we should be expiring the content
pasted to the pasteboard after the same amount of time as
we lock private browsing.

* Source/WebCore/PAL/pal/spi/mac/NSPasteboardSPI.h:
* Source/WebCore/platform/Pasteboard.h:
* Source/WebCore/platform/PlatformPasteboard.h:
* Source/WebCore/platform/ios/PlatformPasteboardIOS.mm:
(WebCore::PlatformPasteboard::setTypes):
(WebCore::PlatformPasteboard::write):
* Source/WebCore/platform/mac/PlatformPasteboardMac.mm:
(WebCore::PlatformPasteboard::write):
(WebCore::PlatformPasteboard::setTypes):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::searchTheWeb):
(WebKit::WebPageProxy::handleContextMenuCopySubject):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::writeSelectionToPasteboard):
(WebKit::WebViewImpl::setPromisedDataForImage):

Canonical link: <a href="https://commits.webkit.org/297124@main">https://commits.webkit.org/297124@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6598b911b37d09e392d60d2878199ca63f1874cb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110507 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30166 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20599 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116533 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60774 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112470 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30845 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38754 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84032 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113455 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24618 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99503 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64473 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23980 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17642 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60328 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93999 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17702 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119323 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37547 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27864 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93004 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37921 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95774 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92827 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23670 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37824 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15565 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33520 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37442 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42913 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37104 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40444 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38813 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b34edad482da458b1144862ce81212191fa3ac65. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47585

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/19eab074d1a899814486ffaf53107850017d570a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47585

whsieh commented on Source/WebCore/platform/Pasteboard.h:
```diff
@@ -72,6 +72,8 @@ class FragmentedSharedBuffer;
 
 struct SimpleRange;
 
+static constexpr unsigned pasteboardExpireTime = 8*60;
```
Nit -
```suggestion
static constexpr auto pasteboardExpirationDelay = 8_min;
```

And then `pasteboardExpirationDelay.seconds()` at the call site when calling `+dateWithTimeIntervalSinceNow:`.

aprotyas commented on Source/WebCore/platform/Pasteboard.h:
```diff
@@ -72,6 +72,8 @@ class FragmentedSharedBuffer;
 
 struct SimpleRange;
 
+static constexpr unsigned pasteboardExpireTime = 8*60;
```
Strong typing may help here (with the explicit time unit)!

```cpp
static constexpr auto pasteboardExpireTime { 8_min };
```

aprotyas replied:
and then you could use it like

```objc
[m_pasteboard _setExpirationDate:[NSDate dateWithTimeIntervalSinceNow:pasteboardExpireTime.seconds()]];
```

aprotyas replied:
also totally optional language nit, but I would call it `duration` over `time` --> `pasteboardExpiryDuration`?

whsieh commented on Source/WebCore/PAL/pal/spi/mac/NSPasteboardSPI.h:
```diff
@@ -34,4 +34,5 @@ extern NSString *_NXSmartPaste;
 
 @interface NSPasteboard ()
 - (NSData *)_dataWithoutConversionForType:(NSString *)type securityScoped:(BOOL)securityScoped;
+- (BOOL)_setExpirationDate:(NSDate*)date;
```
```suggestion
- (BOOL)_setExpirationDate:(NSDate *)date;
```

aprotyas reviewed (commented):
It might be possible to write an API test with an ephemeral web view and some swizzling to confuse the system time (just past the expiration)!

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6598b911b37d09e392d60d2878199ca63f1874cb. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47585

webkit-commit-queue commented:
Commit message contains (OOPS!) and no valid reviewer found, blocking PR #47585. Details: [Build #23649](https://ews-build.webkit.org/#/builders/19/builds/23649)

megangardner committed:
Pasteboard history should not include items copied from ephemeral sessions.
https://bugs.webkit.org/show_bug.cgi?id=295442
rdar://153315024

Reviewed by Abrar Rahman Protyasha.

In order to keep items copied in an ephemeral
session from showing up in pasteboard history,
we need to set a value before writing to the pasteboard.
We also have decided that we should be expiring the content
pasted to the pasteboard after the same amount of time as
we lock private browsing.

* Source/WebCore/PAL/pal/spi/mac/NSPasteboardSPI.h:
* Source/WebCore/platform/Pasteboard.h:
* Source/WebCore/platform/PlatformPasteboard.h:
* Source/WebCore/platform/ios/PlatformPasteboardIOS.mm:
(WebCore::PlatformPasteboard::setTypes):
(WebCore::PlatformPasteboard::write):
* Source/WebCore/platform/mac/PlatformPasteboardMac.mm:
(WebCore::PlatformPasteboard::write):
(WebCore::PlatformPasteboard::setTypes):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::searchTheWeb):
(WebKit::WebPageProxy::handleContextMenuCopySubject):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::writeSelectionToPasteboard):
(WebKit::WebViewImpl::setPromisedDataForImage):

Canonical link: https://commits.webkit.org/297124@main
